asynciterator.ts

```diff
@@ -777,9 +789,95 @@ export class IntegerIterator extends AsyncIterator<number> {
   }
 }
 
+/**
+ * A synchronous mapping function from one element to another.
+ * The iterator performing the mapping is passed as a second argument.
+ */
+export type MapFunction<S, D = S> = (item: S) => D | null;
+
+/**
+ An iterator that calls a synchronous mapping function
+ on every item from its source iterator.
+ @extends module:asynciterator.AsyncIterator
+*/
+export class MappingIterator<S, D = S> extends AsyncIterator<D> {
+  private readonly _destroySource: boolean;
+  protected _source: InternalSource<S>;
+
+  /**
+   * Applies the given mapping to the source iterator.
+   */
+  constructor(
+    _source: AsyncIterator<S>,
+    private _map: MapFunction<S, D>,
+    options: SourcedIteratorOptions = {}
+  ) {
+    super();
+    this._source = _validateSource(_source);
+    this._destroySource = options.destroySource !== false;
+
+    if (_source.done) {
+      this.close();
+    }
+    else {
+      this._source._destination = this;
+      this._source.on('end', destinationClose);
+      this._source.on('error', destinationEmitError);
+      this._source.on('readable', destinationSetReadable);
+      // If we are given a source that is already readable
+      // then we need to set the state of this iterable to readable
+      // also as there is no guarantee that the is no forthcoming
+      // readable event from the source
+      this.readable = this._source.readable;
+    }
+  }
+
+  read(): D | null {
+    // Do not read the source if the current iterator is ended
+    if (this.done)
+      return null;
+
+    // Do not read from the source if it is ended
+    // TODO: Possibly remove this after https://github.com/RubenVerborgh/AsyncIterator/pull/45 is merged
+    // the requirement for this to be removed is that the 'end' emit occurs synchronously when the ENDED
+    // state is set, which does not currently occur if endAsync is called.
+    if (this._source.done) {
+      this.close();
+      return null;
+    }
+
+    let item: D | S | null;
+    while ((item = this._source.read()) !== null) {
+      if ((item = this._map(item)) !== null)
+        return item;
+    }
+
+    // This will set readable to false on the current iterator
+    // if the source is no longer readable
+    this.readable = this._source.readable;
```

Comment:
Right. That's sensible actually.
I'd have been okay with this method triggering another time, but this makes sense.

Reply:
> In fact, if the source returns null, we know that it is not readable

I'm cautious about making that assumption, especially since it doesn't really yield that much in terms of performance gains. For instance, what if someone chose to implement a `FilterIterator` where the `.read` method was

```
read() {
  / * ... readability checks */

  const item = this.source.read()
  return this.filter(item) ? item : null;
}
```
then the source would still be readable.

Reply:
Moreover I can't find anything that makes that assumption explicitly in the [readable API](https://nodejs.org/api/stream.html#event-readable). It makes the statement that "If no data is available to be read, null is returned", but does not say that a `null` return implies it is not readable.

Reply:
I wouldn’t be surprised if a filter method like that were to subtly break things in some use cases. I do think there’s an implicit assumption that a stream that reads null is also a stream that is not readable at that point in time. A safer implementation would read until the source reads null or until the source reads an item that matches the filter.

Reply:
Also, consider this paragraph from https://nodejs.org/api/stream.html#readablereadsize :

> When reading a large file `.read()` may return `null`, having consumed all buffered content so far, but there is still more data to come not yet buffered. In this case a new 'readable' event will be emitted when there is more data in the buffer.

AFAIU, this implies that whenever we `read()` the value `null` we can stop reading and resume doing so upon a new `readable` event.

Reply:
I've been convinced, change pushed